✨ New probes: code-review

andrelmbackman opened the pull request:
#### What kind of change does this PR introduce?

Adds three probes that extends the Code-review check. The probes:
##### codeApproved
 checks if each changeset has been approved at least once
##### codeReviewed
 checks if each changeset has been reviewed at least once
##### codeReviewedTwoReviewers
 checks if each changeset has been reviewed by two unique reviewers


- [✅ ] PR title follows the guidelines defined in our [pull request documentation](https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process)

#### What is the current behavior?

#### What is the new behavior (if this is a feature change)?**

- [✅ ] Tests for the changes have been added (for bug fixes/features)

#### Which issue(s) this PR fixes
Fixes https://github.com/ossf/scorecard/issues/3265
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.

Fixes #

or

NONE
-->

#### Special notes for your reviewer

#### Does this PR introduce a user-facing change?

For user-facing changes, please add a concise, human-readable release note to
the `release-note`

(In particular, describe what changes users might need to make in their
application as a result of this pull request.)

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below.
If the PR requires additional action from users switching to the new release,
include the string "ACTION REQUIRED".

For more information on release notes see: https://git.k8s.io/release/cmd/release-notes/README.md
-->

```release-note
NONE
```

spencerschrock commented on tools/go.mod:
```diff
@@ -45,7 +46,7 @@ require (
 	github.com/Azure/go-autorest/tracing v0.6.0 // indirect
 	github.com/AzureAD/microsoft-authentication-library-for-go v1.0.0 // indirect
-	github.com/BurntSushi/toml v1.3.2 // indirect
+  github.com/BurntSushi/toml v1.3.2 // indirect
```
these `tools/go.{mod,sum}` changes look unintentional. let's remove them from the PR

andrelmbackman replied:
They are indeed, I was resolving a merge conflict in haste.

spencerschrock commented on probes/utils/code_review.go:
```diff
@@ -0,0 +1,51 @@
+)
+
+// returns the number of unique reviewers that aren't the changeset author
+func UniqueReviewers(authorLogin string, reviewers []string) int {
```
all of these functions are only called by one probe. do they need to be in this util folder, instead of part of the the probe that needs it?

andrelmbackman replied:
refactored the probe now, deleted the unnecessary utils

spencerschrock reviewed (dismissed):
Is this part of the conversion process @laurentsimon ? 

I'm going to block on license until we hear from someone definitive.

andrelmbackman commented on probes/codeApproved/def.yml:
```diff
@@ -0,0 +1,20 @@
+# © 2023 Nokia
+# Licensed under the BSD 3 Clause license
+# SPDX-License-Identifier: BSD-3-Clause
```
You are right, the information I had was outdated! Changing to Apache 2.

andrelmbackman commented on probes/entries.go:
```diff
@@ -30,19 +33,25 @@ var (
 		toolPyUpInstalled.Run,
 		toolSonatypeLiftInstalled.Run,
 	}
+	CodeReviewChecks = []ProbeImpl {
```
Sgtm

naveensrinivasan reviewed (changes requested):
Thanks, Please ensure all the code are covered with tests. 

We are in the process of having high code coverage.

raghavkaul reviewed (changes requested):
Thank you for the contribution! Share Naveen's thoughts that this needs more unit tests.

andrelmbackman commented on probes/codeReviewTwoReviewers/impl.go:
```diff
@@ -0,0 +1,110 @@
+
+const probe = "codeReviewTwoReviewers"
+const minimumReviewers = 2
+const noReviewerFound = -1
```
Will do, that is the pattern I should get used to. Bear with me as I transition from my C-style syntax!

andrelmbackman commented:
> Thanks, Please ensure all the code are covered with tests.
> 
> We are in the process of having high code coverage.

I am making tests akin to the tests of the dependency_update_tool check, which uses probes. In this case, I will have to refactor checks/code_review.go and checks/evaluation/code_review.go, right? Not sure how much the outcomes of each probe should weigh in that case, suggestions?

spencerschrock commented on probes/codeApproved/def.yml:
```diff
@@ -0,0 +1,21 @@
+# © 2023 Nokia
+# Licensed under the Apache License 2.0
+# SPDX-License-Identifier: Apache-2.0
```
Thanks for changing this to Apache 2.0

Can we keep the license text similar to what we have in the other files?

```
// Copyright 2023 OpenSSF Scorecard Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
```

spencerschrock replied:
Note: I did ask about listing `Nokia` and was told that it's a valid option. 

```
// Copyright 2023 Nokia
```

When someone else edits the file we will need to change it to include other authors as well. Which is partly why we use "OpenSSF Scorecard Authors" now as a catchall, so we don't need to go and update it the next time someone updates this file. Nokia would still own the copyright of your contribution.

andrelmbackman replied:
I checked with my open source specialist, and it's ok, I can change it to the suggested license text.

spencerschrock reviewed (commented):
Spoke with the Linux Foundation today. Going to clear my previously blocking review.

raghavkaul commented:
@andrelmbackman One way to unit test is to declare and pass your own `RawResults` or `CodeReviewData` (this is how the existing unit tests work) - neither of which would require refactoring the existing checks.

andrelmbackman commented on probes/codeReviewed/impl.go:
```diff
@@ -0,0 +1,58 @@
+	for x := range changesets {
+		data := &changesets[x]
+		if len(data.Reviews) > 0 {
+			numReviews += 1
```
Just checking if each changeset has any number of reviews, only increments once for each changeset.

andrelmbackman commented on probes/codeReviewTwoReviewers/impl.go:
```diff
@@ -0,0 +1,110 @@
+	return uniqueReviewers
+}
+
+func authorNotFound(findings []finding.Finding, probeID string,
```
Changed it now

andrelmbackman commented on probes/codeReviewTwoReviewers/impl.go:
```diff
@@ -0,0 +1,110 @@
+	return findings, probeID, nil
+}
+
+func uniqueReviewers(authorLogin string, reviewers []string) int {
```
Implemented a simpler helper function like you suggested, commented it and added test in ./checks/code_review_test.go

raghavkaul commented on probes/codeReviewTwoReviewers/impl.go:
```diff
@@ -0,0 +1,125 @@
+	return codeReviewRun(rawReviewData, fs, probe, finding.OutcomePositive, finding.OutcomeNegative)
+}
+
+/*
```
Nit: use `//` style comments for consistency with the rest of the code.

raghavkaul commented on probes/utils/codeReview.go:
Nit: `utils/code_review.go`

raghavkaul commented on probes/entries.go:
```diff
@@ -12,11 +12,15 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//nolint // to prevent linter errors because of package names being camel case
```
Nit: determine the particular `golangci-lint` check that fails, and rewrite this nolint as `nolint:<checkname>`.

andrelmbackman replied:
fixed

raghavkaul commented on probes/codeReviewTwoReviewers/impl.go:
```diff
@@ -0,0 +1,125 @@
+		return findings, probeID, nil
+	} else if leastFoundReviewers >= minimumReviewers {
+		// returns PositiveOutcome if the lowest number of unique reviewers is at least as high as minimumReviewers (2).
+		f, err := finding.NewWith(fs, probeID, fmt.Sprintf("%v unique reviewers found for at least one " +
```
Should this message be phrased as ">%d unique reviewers found for all changesets"?

raghavkaul commented on probes/codeReviewTwoReviewers/impl.go:
```diff
@@ -0,0 +1,125 @@
+		findings = append(findings, *f)
+	} else {
+		// returns NegativeOutcome if even a single changeset was reviewed by fewer than minimumReviewers (2).
+		f, err := finding.NewWith(fs, probeID, fmt.Sprintf("%v unique reviewer(s) found for at least one " +
```
Should this message be phrased as "some changesets had <%d unique reviewers"?

raghavkaul commented on checks/code_review_test.go:
```diff
@@ -275,6 +282,247 @@ func TestCodereview(t *testing.T) {
 			},
 		},
 	}
+	probeTests := []struct {
```
Could this be moved to a separate test function for clarity/maintainability?

andrelmbackman replied:
Done

raghavkaul commented on probes/codeApproved/impl.go:
```diff
@@ -0,0 +1,61 @@
+func approvedRun(reviewData *checker.CodeReviewData, fs embed.FS, probeID string,
+	positiveOutcome, negativeOutcome finding.Outcome,
+) ([]finding.Finding, string, error) {
+	var findings []finding.Finding
```
If all but one changeset is bot authored, the entire review score would come down to the human-authored changeset and any unapproved bot changesets, to maintain consistency with `checks/evaluation/code_review.go:L54`

raghavkaul reviewed (commented):
Thanks for addressing the previous review! Some small stylistic suggestions, and if you could address the other open comment threads on GitHub/resolve if they're not relevant, that would be what we need to move forward.

laurentsimon commented on checks/code_review.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2020 OpenSSF Scorecard Authors
+// Copyright 2023 OpenSSF Scorecard Authors
```
you need to update this file to call the probe, see https://github.com/ossf/scorecard/blob/main/checks/security_policy.go#L47-L59

That will require updating the check/evaluation/core_review.go as well

laurentsimon replied:
See https://github.com/ossf/scorecard/pull/3230 as an example PR

andrelmbackman replied:
I asked about it earlier but did not get a clear answer, thought these probes would only be experimental for now. Having said that, I am more than willing to update the evaluation! I need some opinions on how to score the results of the probes:
`codeApproved
codeReviewed
codeReviewTwoReviewers`?
The obvious results being:
- if all return OutcomePositive: CreateMaxScoreResult
- if all return OutcomeNegative: CreateMinScoreResult.

What about if codeApproved and codeReviewed pass, but codeReviewTwoReviewers fails, how much should they weigh when calculating the final score? How about OutcomeNotAvailable? Thankful for any input on this.

raghavkaul replied:
Is this required, or can we keep both the probes and the scoring changes separate and updated in parallel?

andrelmbackman commented:
docker-build fails for some reason unknown to me, is there a way to trigger it again?

```
> [base 4/5] RUN go mod download:
22.30 go: golang.org/x/text@v0.11.0: read "https:/proxy.golang.org/@v/v0.11.0.zip": stream error: stream ID 2437; INTERNAL_ERROR; received from peer

Dockerfile:19

  17 |     ENV CGO_ENABLED=0
  18 |     COPY go.* ./
  19 | >>> RUN go mod download
  20 |     COPY . ./
  21 |     

ERROR: failed to solve: process "/bin/sh -c go mod download" did not complete successfully: exit code: 1
```

andrelmbackman commented:
Assuming that these probe implementations will be used in the final evaluation for the score, how would you suggest I distribute the scoring? Would it be a good distribution if codeApproved weighs 4 points, codeReviewed weighs 3 points and codeReviewTwoReviewers weighs 3 points, assuming OutcomePositive? OutcomeNotAvailable could score 1 point.

laurentsimon commented:
> Assuming that these probe implementations will be used in the final evaluation for the score, how would you suggest I distribute the scoring? Would it be a good distribution if codeApproved weighs 4 points, codeReviewed weighs 3 points and codeReviewTwoReviewers weighs 3 points, assuming OutcomePositive? OutcomeNotAvailable could score 1 point.

can you remind us the difference between codeApproved and codeReviewed?

andrelmbackman commented:
> > Assuming that these probe implementations will be used in the final evaluation for the score, how would you suggest I distribute the scoring? Would it be a good distribution if codeApproved weighs 4 points, codeReviewed weighs 3 points and codeReviewTwoReviewers weighs 3 points, assuming OutcomePositive? OutcomeNotAvailable could score 1 point.
> 
> can you remind us the difference between codeApproved and codeReviewed?

- both return OutcomeNotAvailable if all changesets were bot-authored
- codeApproved returns OutcomePositive if all changesets have at least one review with the state "APPROVED"
- codeReviewed returns OutcomePositive if all changesets have review activity (if there are reviews in each DefaultBranchChangeset)

laurentsimon commented:
> > 
> > can you remind us the difference between codeApproved and codeReviewed?
> 
> * both return OutcomeNotAvailable if all changesets were bot-authored

should they return failure instead?

> * codeApproved returns OutcomePositive if all changesets have at least one review with the state "APPROVED"
> * codeReviewed returns OutcomePositive if all changesets have review activity (if there are reviews in each DefaultBranchChangeset)

sorry, I still don't understand what codeReviewed does. What state are we looking for in each changset? How many commits in the changeset are required to have the state?

raghavkaul commented:
> > > can you remind us the difference between codeApproved and codeReviewed?
> > 
> > 
> > 
> > * both return OutcomeNotAvailable if all changesets were bot-authored
> 
> should they return failure instead?
> 
I would reserve Failure for cases where there's an SCM system that probably implements a review/approval system but we don't know how to check for it, and NotApplicable for cases where there's insufficient data, no human commits, etc. But it's not a strong preference/blocker for me.

> > * codeApproved returns OutcomePositive if all changesets have at least one review with the state "APPROVED"
> > * codeReviewed returns OutcomePositive if all changesets have review activity (if there are reviews in each DefaultBranchChangeset)
> 
> sorry, I still don't understand what codeReviewed does. What state are we looking for in each changset? How many commits in the changeset are required to have the state?

From my understanding, codeReviewed is any review activity whatsoever, like a comment, even if it's short of an explicit approval.

We associate reviews and approvals by changeset, not commit, so any one of the commits in a changeset can be approved for the entire changeset to be approved. The difference I think would be between approvals on HEAD vs merges with unreviewed commits, but I think the branch protection check handles that better than the code review check.

laurentsimon commented:
> > > > can you remind us the difference between codeApproved and codeReviewed?
> > > 
> > > 
> > > 
> > > * both return OutcomeNotAvailable if all changesets were bot-authored
> > 
> > 
> > should they return failure instead?
> 
> I would reserve Failure for cases where there's an SCM system that probably implements a review/approval system but we don't know how to check for it, and NotApplicable for cases where there's insufficient data, no human commits, etc. But it's not a strong preference/blocker for me.

gotcha. Thanks

> > > * codeApproved returns OutcomePositive if all changesets have at least one review with the state "APPROVED"
> > > * codeReviewed returns OutcomePositive if all changesets have review activity (if there are reviews in each DefaultBranchChangeset)
> > 
> > 
> > sorry, I still don't understand what codeReviewed does. What state are we looking for in each changset? How many commits in the changeset are required to have the state?
> 
> From my understanding, codeReviewed is any review activity whatsoever, like a comment, even if it's short of an explicit approval.

Is this actionable / useful?

Why not a "the last commit got an 'APPROVED'"? (That would match on of the branch protection settings that forces a new review if there is a new commit)

> We associate reviews and approvals by changeset, not commit, so any one of the commits in a changeset can be approved for the entire changeset to be approved. The difference I think would be between approvals on HEAD vs merges with unreviewed commits, but I think the branch protection check handles that better than the code review check.

yes BP has it, but you could argue it's point in time...

In terms of levels / hardening:
1. all PRs reviews
2. all PRs are twice reviewed.
3. Last commit is always reviewed
4. last commit is always reviewed twice (I don't think this will ever happen, so (3) is probably enough)

andrelmbackman commented:
> * Last commit is always reviewed
> * last commit is always reviewed twice (I don't think this will ever happen, so (3) is probably enough)

Thank you for your input! The reviews are not referencing specific commits, there is no association in the raw results. The probes are checking the slice of reviews for a changeset, and in the case of codeApproved, returns positiveOutcome if one of them has the state set to 'approved'. Currently, the scorecard only checks if there is review activity in the changeset, and checks for the 'approved' state only if the platform is GitHub, otherwise it gives full points. Heads-up: my contract (and access to the fork) ends this week on Friday.

raghavkaul commented:
> Is this actionable / useful?
> 
> Why not a "the last commit got an 'APPROVED'"? (That would match on of the branch protection settings that forces a new review if there is a new commit)
> 
I don't think it's a super actionable probe, in fact, it might make people think that some discussion, no matter how low-quality, constitutes good code review.

> In terms of levels / hardening:
>  ...
> 
Agree with the prioritization. So, can we replace the `codeReviewed` probe maybe with something like `code(Reviewed|Approved)AtHead`? Or, unless there's a specific use case that you have in mind @andrelmbackman, I'm inclined to agree that that probe is unnecessary and could simply be removed, and a follow-up could add the other probe.

andrelmbackman commented:
Removed the codeReviewed probe, but the evaluation part needs work. I think there is a need for more probes, e.g. validating bot commits. The scoring would differ so much from the current state with just these two probes that I don't think that Code-Review check should be migrated to support probes just yet. As stated earlier, today is my last day of working on the Nokia fork, so further work on this PR will not be possible. Having said that, I can help in the future (from my own fork) with migrating the check to support probes if needed. Is it possible that these probes can exist just in the experimental part for now?

p.s. Moved the probe tests to their own package directories, CodeCov is still not happy.

raghavkaul reviewed (approved):
Thanks, this looks good to me. I'm somewhat unclear on whether we should have `codeReviewTwoReviewers` or make the condition on `probes/codeReviewTwoReviewers/impl.go:L123` require that the review state is `APPROVED`, similar to the other probe, but I don't have a problem with the probe as is. 

As for scoring, I don't think we should modify `Code-Review` checks' scoring, but rather keep the probes updated in parallel.

jitsengupta17 commented:
Hi All Reviewers,

we need to open this branch, for further code commits.
andrelmbackman was earlier part of the organization now i will be contributing in his place.
please open the closed branch.

CsatariGergely commented:
@azeemshaikh38 @justaugustus @laurentsimon can you please reopen this PR, so @jitsengupta17 can continue the work on it?

spencerschrock commented:
Whoops, slipped through the cracks. Thanks for the followup ping.

spencerschrock commented:
Note: We've been working on more probe conversions in the meanwhile, so some of the semantics may have changed since time of last review. So there may be additional feedback on the way

jitsengupta17 commented:
@spencerschrock : we have resolved the conflicts, if any additional feedback please let us know. 
we are going to rename the file codeReviewedTwoReviewers to codeReviewedOneReviewers : now the code will required one additional reviewer other than the author to review.

spencerschrock commented:
> @spencerschrock : we have resolved the conflicts, if any additional feedback please let us know. we are going to rename the file codeReviewedTwoReviewers to codeReviewedOneReviewers : now the code will required one additional reviewer other than the author to review.

@raghavkaul will re-review

jitsengupta17 commented:
@raghavkaul : thanks for approving my previous conflicts,
please review and approve this conflict as-well.

raghavkaul commented on checks/code_review_test.go:
All changes in this file seem unrelated to the PR?

raghavkaul commented on probes/codeApproved/def.yml:
```diff
@@ -0,0 +1,31 @@
+  To ensure that the review process works, the proposed changes
+  should have a minimum number of approvals.
+implementation: >
+  The implementation looks for the reviews of a changeset and that their state is "APPROVED"
```
```suggestion
  This probe looks for whether all changes over the last `--commit-depth` commits have been approved by a maintainer. Commits are grouped by the Pull Request they were introduced in, and each Pull Request must have at least one approval.
```

raghavkaul commented on probes/codeApproved/def.yml:
```diff
@@ -0,0 +1,31 @@
+implementation: >
+  The implementation looks for the reviews of a changeset and that their state is "APPROVED"
+outcome:
+  - If all changesets were approved, the probe returns OutcomePositive (1)
+  - If a changeset was not approved, the prove returns OutcomeNegative (0)
```
```suggestion
  - If all commits were approved, the probe returns OutcomePositive (1)
  - If a commits was not approved, the prove returns OutcomeNegative (0)
```

raghavkaul commented on probes/codeApproved/def.yml:
```diff
@@ -0,0 +1,31 @@
+remediation:
+  effort: Low
+  text:
+    - Follow the instructions https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews.
```
```suggestion
    - Follow the instructions at https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews.
```

raghavkaul commented on probes/codeApproved/def.yml:
```diff
@@ -0,0 +1,31 @@
+  effort: Low
+  text:
+    - Follow the instructions https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews.
+  markdown:
+    - Follow the instructions https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews.
```
```suggestion
  markdown:
    - Follow the instructions [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews) to review pull requests before merge.
```

raghavkaul commented on probes/codeReviewTwoReviewers/impl.go:
```diff
@@ -0,0 +1,128 @@
+			}
+			findings = append(findings, *f)
+			return findings, probeID, nil
+		} else if !data.Author.IsBot {
```
Would be cleaner if we did the same thing as in the other probe L63-65, `if` instead of `else if`

jitsengupta17 replied:
done in the new file probes/codeReviewOneReviewers/impl.go

raghavkaul commented on probes/codeApproved/impl.go:
```diff
@@ -0,0 +1,98 @@
+		return nil, probeID, utils.NoChangesetsErr
+	}
+	for x := range changesets {
+		data := &changesets[x]
```
Let's add checks for:
1. `data.Author.Login != ''`
2. `data.Reviews[y].Author.Login != ''`

Just like in the other probe. If `Login == ''` then we should match the other probe and return `finding.NewNotAvailable`.

jitsengupta17 replied:
done

raghavkaul reviewed (changes requested):
Overall looks good, thanks for adding tests! Some code layout and error handling comments.

jitsengupta17 commented:
I have resolved some conflicts, and renamed the file codeReviewedTwoReviewers to codeReviewedOneReviewers : now the code will required one additional reviewer other than the author to review.
@raghavkaul : please review the changes

raghavkaul commented on probes/codeApproved/def.yml:
```diff
@@ -0,0 +1,31 @@
+  To ensure that the review process works, the proposed changes
+  should have a minimum number of approvals.
+implementation: >
+  This probe looks for whether all changes over the last `--commit-depth` commits have been approved by a maintainer. Commits are grouped by the Pull Request they were introduced in, and each Pull Request must have at least one approval.
```
```suggestion
  This probe looks for whether all changes over the last `--commit-depth` commits have been approved before merge. Commits are grouped by the Pull Request they were introduced in, and each Pull Request must have at least one approval.
```

raghavkaul commented on probes/codeReviewOneReviewers/def.yml:
```diff
@@ -0,0 +1,31 @@
+  To ensure that the review process works, the proposed changes
+  should have a minimum number of approvals.
+implementation: >
+  The implementation looks for the number of reviewers for the last few changesets.
```
```suggestion
This probe looks for whether all changes over the last `--commit-depth` commits have been approved by a minimum number of reviewers.
Commits are grouped by the Pull Request they were introduced in.
```

raghavkaul reviewed (changes requested):
Looks good, thank you for all the work! Small changes to documentation and we'll be good to go.

jitsengupta17 commented:
> Looks good, thank you for all the work! Small changes to documentation and we'll be good to go.

raghavkaul commented:
@jitsengupta17 @gowriNSN You'll need to ensure the tests pass. Looks like mostly linter/format issues.

jitsengupta17 commented:
@raghavkaul : can you please review, fixed all linting and test cases.

jitsengupta17 commented:
> @jitsengupta17 @gowriNSN You'll need to ensure the tests pass. Looks like mostly linter/format issues.

please approve, fixed all linting and test cases.

jitsengupta17 commented:
> @jitsengupta17 @gowriNSN You'll need to ensure the tests pass. Looks like mostly linter/format issues.

@raghavkaul : please review, fixed all linting and test cases. please Approve

raghavkaul commented:
@jitsengupta17 There are still some lint errors in the latest run.

raghavkaul commented:
@jitsengupta17 Ping on the linter issues!

jitsengupta17 commented:
@raghavkaul : I have fix some of the linter issues but i am still getting 2 in our local run. The below errors :

checks/code_review_test.go:36: File is not `gci`-ed with --skip-generated -s standard -s default -s prefix(github.com/ossf/scorecard) (gci)

make: *** [Makefile:95: check-linter] Error 1

I am not able to find the reason for the above two error. can you please help. the other fix you can review

raghavkaul commented on probes/entries.go:
```diff
@@ -12,13 +12,18 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+ //nolint:gci
+ //nolint:gofmt
+ //nolint:gofumpt
```
```suggestion
```

raghavkaul commented on checks/code_review_test.go:
```diff
@@ -27,9 +27,15 @@ import (
 // TestCodeReview tests the code review checker.
 func TestCodereview(t *testing.T) {
 	t.Parallel()
+	// fieldalignment lint issue. Ignoring it as it is not important for this test.
+	//nolint:gci
+    //nolint:gofmt
+    //nolint:gofumpt
```
```suggestion
```

raghavkaul reviewed (commented):
I think these `nolint`s are still throwing off `make check-linter`.

jitsengupta17 commented:
@raghavkaul : i missed one //nolint:goimports . This time it should work. please review

jitsengupta17 commented:
@raghavkaul : please review i have removed the linters from probes/entries.go

jitsengupta17 commented:
@raghavkaul : Thanks for reviewing my previous code ,the error are now removed please review this changes

raghavkaul commented:
@jitsengupta17 Would you like to give me write access to your repository/branch so I can help get this ready to merge?

jitsengupta17 commented:
@raghavkaul: access right has been provided to you. please help us to merge the code. Thanks

CsatariGergely commented:
> @jitsengupta17 Would you like to give me write access to your repository/branch so I can help get this ready to merge?

@raghavkaul I've sent an invite for you. After you accept the invitation you will be able to push to nokia:probes/code-review. Thanks.

jitsengupta17 commented:
@raghavkaul : Thank you so much for helping us to merge the code.
